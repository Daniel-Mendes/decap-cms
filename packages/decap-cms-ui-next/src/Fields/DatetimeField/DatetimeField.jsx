import React, { useEffect, useState } from 'react';
import styled from '@emotion/styled';
import dayjs from 'dayjs';
import { DatetimePicker } from 'rc-datetime-picker-dayjs';

import TextField from '../TextField';
import { Menu } from '../../Menu';
import DatetimePickerStyles from './DatetimePickerStyles';

const StyledMenu = styled(Menu)`
  padding: 0;
`;

const StyledDatetimePicker = styled(DatetimePicker)`
  width: 100%;
  background-color: transparent;
`;

function DatetimeField({ onChange, value, type, shortcuts, readOnly, ...props }) {
  const [date, setDate] = useState(value);
  const [inputFormat, setInputFormat] = useState('YYYY-MM-DDTHH:mm:ssZ');
  const [anchorEl, setAnchorEl] = useState(null);

  useEffect(() => {
    if (type === 'date') {
      setInputFormat('YYYY-MM-DD');
    } else if (type === 'time') {
      setInputFormat('HH:mm');
    } else {
      setInputFormat('YYYY-MM-DDTHH:mm');
    }
  }, [type]);

  function handleOpenMenu(event) {
    // setAnchorElWidth(`${event.currentTarget.offsetWidth}px`);
    setAnchorEl(event.currentTarget);
  }

  function handleClose(value) {
    if (value && typeof value === 'string') {
      onChange(value);
    }
    setAnchorEl(null);
  }

  return (
    <>
      <DatetimePickerStyles shortcuts={shortcuts} />

      <TextField
        {...props}
        readOnly
        value={value}
        type={type}
        onClick={!readOnly && handleOpenMenu}
        focus={!!anchorEl}
        icon={type === 'time' ? 'clock' : 'calendar'}
      />

      <StyledMenu
        anchorOrigin={{ x: 'right', y: 'bottom' }}
        transformOrigin={{ x: 'right', y: 'top' }}
        anchorEl={anchorEl}
        open={!!anchorEl}
        onClose={handleClose}
        width={'300px'}
      >
        <StyledDatetimePicker
          shortcuts={shortcuts}
          showCalendarPicker={type === 'date' || type === 'datetime-local'}
          showTimePicker={type === 'time' || type === 'datetime-local'}
          closeOnSelectDay
          dayjs={date}
          onChange={date => {
            setDate(date);
            onChange(date);
          }}
        />
      </StyledMenu>
    </>
  );
}
export default DatetimeField;
