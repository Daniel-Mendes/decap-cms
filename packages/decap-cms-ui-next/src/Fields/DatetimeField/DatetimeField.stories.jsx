import React from 'react';
import styled from '@emotion/styled';
import { useArgs } from '@storybook/preview-api';
import dayjs from 'dayjs';

import DatetimeField from '.';

const StyledDatetimeField = styled(DatetimeField)`
  width: 33vw;
`;

export default {
  title: 'Fields/DatetimeField',
  component: DatetimeField,
  argTypes: {
    value: {
      control: 'date',
    },
    type: {
      control: 'select',
      options: ['date', 'time', 'datetime-local'],
    },
    shortcuts: {
      control: 'object',
    },
    errors: {
      control: 'object',
      if: {
        arg: 'error',
      },
    },
  },
  args: {
    label: 'Datetime Field Label',
    status: 'Required',
    description: 'This is a description',
    value: new Date(),
    type: 'datetime-local',
    shortcuts: {
      Now: dayjs(),
      Clear: '',
    },
    inline: false,
    error: false,
    errors: [
      {
        message: 'Error message.',
      },
    ],
  },
};

export function _DatetimeField(args) {
  const [{ value }, updateArgs] = useArgs();

  function handleValue(date) {
    updateArgs({ value: date });
  }

  return <StyledDatetimeField {...args} value={dayjs(value)} onChange={handleValue} />;
}
