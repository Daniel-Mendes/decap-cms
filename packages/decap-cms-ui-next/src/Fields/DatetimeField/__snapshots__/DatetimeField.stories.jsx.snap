// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fields/DatetimeField _DatetimeField smoke-test 1`] = `"<div class="e2uy6hg0 css-m616ke-FieldWrap-StyledDatetimeField e1x0ipyd6"><div class="css-kuth3m-FieldInside-FieldInside-FieldInside-FieldInside e1x0ipyd5"><label class="css-1n9k11u-Label-StyledLabel e1x0ipyd3">Datetime Field Label <div color="neutral" radius="large" class="css-6930ng-BadgeWrap-BadgeWrap epfx4ki3">Required</div></label><div class="css-dj548v-ChildrenWrap e1x0ipyd4"><input readonly="" type="datetime-local" placeholder="Type datetime field label here" autocomplete="off" class="css-gfgrnx-StyledInput e4jt8f0" value="2024-08-17T15:07"></div><button class="e1x0ipyd2 css-12u9ugq-IconButtonWrap-StyledIconButton e1s566520"><svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" class="" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"><rect x="3" y="4" width="18" height="18" rx="2" ry="2"></rect><line x1="16" y1="2" x2="16" y2="6"></line><line x1="8" y1="2" x2="8" y2="6"></line><line x1="3" y1="10" x2="21" y2="10"></line></svg></button><p class="css-nalsj0-StyledDescription e1x0ipyd1">This is a description</p></div></div>"`;
